src/backend/gslice.c

```diff
@@ -0,0 +1,256 @@
+#include        "global.h"
+#include        "type.h"
+#include        "code.h"
+
+static char __file__[] = __FILE__;      /* for tassert.h                */
+#include        "tassert.h"
+
+struct SymInfo
+{
+    bool canSlice;
+    SYMIDX si0;
+};
+
+static void sliceStructs_Gather(SymInfo *sia, elem *e)
+{
+    while (1)
+    {
+        switch (e->Eoper)
+        {
+            case OPvar:
+            {
+                SYMIDX si = e->EV.sp.Vsym->Ssymnum;
+                if (si >= 0 && sia[si].canSlice)
+                {
+                    assert(si < globsym.top);
+                    if (tysize(e->Ety) == REGSIZE &&
+                        (e->Eoffset == 0 || e->Eoffset == REGSIZE))
+                    {
+                    }
+                    else
+                    {
+                        sia[si].canSlice = false;
+                    }
+                }
+                return;
+            }
+            default:
+                if (OTunary(e->Eoper))
+                {
+                    e = e->E1;
+                    break;
+                }
+                if (OTbinary(e->Eoper))
+                {
+                    sliceStructs_Gather(sia, e->E2);
+                    e = e->E1;
+                    break;
+                }
+                return;
+        }
+    }
+}
+
+static void sliceStructs_Replace(SymInfo *sia, elem *e)
+{
+    while (1)
+    {
+        switch (e->Eoper)
+        {
+            case OPvar:
+            {
+                Symbol *s = e->EV.sp.Vsym;
+                SYMIDX si = s->Ssymnum;
+                //printf("e: %d %d\n", si, sia[si].canSlice);
+                //elem_print(e);
+                if (si >= 0 && sia[si].canSlice)
+                {
+                    if (e->Eoffset == 0)
+                    {
+                        type_free(s->Stype);
+                        s->Stype = type_fake(e->Ety);
+                    }
+                    else
+                    {
+                        Symbol *s1 = globsym.tab[sia[si].si0 + 1];
+                        type_free(s1->Stype);
+                        s1->Stype = type_fake(e->Ety);
+                        e->EV.sp.Vsym = s1;
+                        e->Eoffset = 0;
+                        //printf("replaced with:\n");
+                        //elem_print(e);
+                    }
+                }
+                return;
+            }
+
+            default:
+                if (OTunary(e->Eoper))
+                {
+                    e = e->E1;
+                    break;
+                }
+                if (OTbinary(e->Eoper))
+                {
+                    sliceStructs_Replace(sia, e->E2);
+                    e = e->E1;
+                    break;
+                }
+                return;
+        }
+    }
+}
+
+void sliceStructs()
+{
+    if (debugc) printf("sliceStructs()\n");
+    size_t sia_length = globsym.top;
+    SymInfo *sia = (SymInfo *)malloc(3 * sia_length * sizeof(SymInfo));
```

Comment:
3 is because it is used for two arrays, sia[] and sia2[]. sia2[] can grow to twice the size of sia[], as symbols can get split into two.